### PR TITLE
chore: Move all remaining actions into links

### PR DESCRIPTION
**Pull request by sharat87**
The chain+link framework is now used for all the steps in the backup command.

## Automation

/test sanity

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12080116829>
> Commit: fcc016fd1fa0bbe9b4cae2478bf4f7750ee5459d
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12080116829&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Sanity`
> Spec:
> <hr>Fri, 29 Nov 2024 07:12:22 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Release Notes

- **New Features**
	- Introduced new classes for managing backup processes, including `BackupFolderLink`, `DiskSpaceLink`, `EnvFileLink`, `GitStorageLink`, `MongoDumpLink`, and `ManifestLink`.
	- Added functionality for checking available disk space and managing temporary backup folders.
	- Enhanced encryption handling with a new password prompt mechanism.

- **Bug Fixes**
	- Streamlined backup process by removing redundant functions and ensuring proper error handling.

- **Documentation**
	- Updated comments and documentation for new classes and methods to improve clarity.

- **Refactor**
	- Restructured backup functionality for improved modularity and maintainability.

### CODE REVIEW

File: app/client/packages/rts/src/ctl/backup/links/MongoDumpLink.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Exports the MongoDB database data using mongodump.
+ */
+export class MongoDumpLink implements Link {
+  constructor(private readonly state: BackupState) {}
+
+  async doBackup() {
+    console.log("Exporting database");
+    await executeMongoDumpCMD(this.state.backupRootPath, utils.getDburl());
+    console.log("Exporting database done.");
+  }
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/MongoDumpLink.ts:15]: Exceptions are caught outside already. It doesn't make sense to catch the error and then throw it again.

File: app/client/packages/rts/src/ctl/backup/links/MongoDumpLink.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+export async function executeMongoDumpCMD(destFolder: string, dbUrl: string) {
+  return await utils.execCommand([
+    "mongodump",
+    `--uri=${dbUrl}`,
+    `--archive=${destFolder}/mongodb-data.gz`,
+    "--gzip",
+  ]);
+}
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/MongoDumpLink.ts:25]: Exceptions are caught outside already. It doesn't make sense to catch the error and then throw it again.

File: app/client/packages/rts/src/ctl/backup/links/BackupFolderLink.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      path.join(os.tmpdir(), "appsmithctl-backup-"),
+    );
+  }
+
+  async postBackup() {
+    await fsPromises.rm(this.state.backupRootPath, {
+      recursive: true,
+      force: true,
+    });
+  }
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/BackupFolderLink.ts:25]: Exceptions are caught outside already. It doesn't make sense to catch the error and then throw it again.

File: app/client/packages/rts/src/ctl/backup/links/GitStorageLink.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  async doBackup() {
+    console.log("Creating git-storage archive");
+
+    const gitRoot = getGitRoot(process.env.APPSMITH_GIT_ROOT);
+
+    await executeCopyCMD(gitRoot, this.state.backupRootPath);
+    console.log("Created git-storage archive");
+  }
+}
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/GitStorageLink.ts:20]: Exceptions are caught outside already. It doesn't make sense to catch the error and then throw it again.

File: app/client/packages/rts/src/ctl/backup/links/EnvFileLink.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      !line.startsWith("APPSMITH_ENCRYPTION") &&
+      !line.startsWith("APPSMITH_MONGODB") &&
+      !line.startsWith("APPSMITH_DB_URL=")
+    ) {
+      output_lines.push(line);
+    }
+  });
+
+  return output_lines.join("\n");
+}
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/EnvFileLink.ts:65]: Don't want to touch this logic in this PR now. Very sensitive and likely very flaky.

File: app/client/packages/rts/src/ctl/backup/links/EnvFileLink.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        "\nAPPSMITH_ENCRYPTION_PASSWORD=" +
+        process.env.APPSMITH_ENCRYPTION_PASSWORD;
+    }
+
+    await fsPromises.writeFile(
+      this.state.backupRootPath + "/docker.env",
+      cleanedContent,
+    );
+    console.log("Exporting docker environment file done.");
+  }
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/EnvFileLink.ts:41]: No. The location of `docker.env` should NOT be configurable. It should be liberally hard-coded.

File: app/client/packages/rts/src/ctl/backup/links/EncryptionLink.ts [hunk 7]
```diff
@@ -34,3 +43,64 @@ export class EncryptionLink implements Link {
...
+    "-pbkdf2",
+    "-iter",
+    "100000",
+    "-in",
+    archivePath,
+    "-out",
+    encryptedArchivePath,
+    "-k",
+    encryptionPassword,
+  ]);
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/EncryptionLink.ts:103]: This is okay for now. I'll revisit later.

File: app/client/packages/rts/src/ctl/backup/links/BackupFolderLink.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * be assumed in the "doBackup" step, and no other.
+ */
+export class BackupFolderLink implements Link {
+  constructor(private readonly state: BackupState) {}
+
+  async preBackup() {
+    this.state.backupRootPath = await fsPromises.mkdtemp(
+      path.join(os.tmpdir(), "appsmithctl-backup-"),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (sharat87) [reply app/client/packages/rts/src/ctl/backup/links/BackupFolderLink.ts:18]: Exceptions are caught outside already. It doesn't make sense to catch the error and then throw it again.

**Review by mohanarpit — APPROVED** [review]

**Pull request merged**